apps/openmw/mwworld/containerstore.cpp

```diff
@@ -79,13 +79,20 @@ namespace
 
 MWWorld::ResolutionListener::~ResolutionListener()
 {
-    if(!mStore.mModified && mStore.mResolved && !mStore.mPtr.isEmpty())
+    try
     {
-        for(const MWWorld::Ptr& ptr : mStore)
-            ptr.getRefData().setCount(0);
-        mStore.fillNonRandom(mStore.mPtr.get<ESM::Container>()->mBase->mInventory, "", mStore.mSeed);
-        addScripts(mStore, mStore.mPtr.mCell);
-        mStore.mResolved = false;
+        if(!mStore.mModified && mStore.mResolved && !mStore.mPtr.isEmpty())
+        {
+            for(const MWWorld::Ptr& ptr : mStore)
+                ptr.getRefData().setCount(0);
+            mStore.fillNonRandom(mStore.mPtr.get<ESM::Container>()->mBase->mInventory, "", mStore.mSeed);
+            addScripts(mStore, mStore.mPtr.mCell);
+            mStore.mResolved = false;
+        }
+    }
+    catch(const std::runtime_error& e)
+    {
+        Log(Debug::Error) << "Error in the ResolutionListener destructor: " << e.what();
     }
```

Comment:
Are we sure this is necessary?

It seems that the reason anything is 'thrown' is because of this:
apps/openmw/mwworld/containerstore.cpp:
```
    if (ptr.isEmpty())
        throw std::runtime_error ("invalid iterator");
```

Since you are just 'catching' it and logging the error... why not just log the error there instead of throwing it. Then is this try/catch block not necessary.

Reply:
> why not just log the error there instead of throwing it

I do not know much about low-level code, so I can not tell if it is a valid behaviour in any case where iteration over container store is used.

Reply:
It was written by @zinnschlag in 2012, probably with other ideas.  I believe it would be better to just log an Error... or perhaps a warning, "invalid iterator" if the ptr is empty and not store it in the container. That way we know something is wrong and an go looking... that being said, in the years since 2012... we've never seen that runtime_error pop up.

Reply:
BTW, it is an operator*(), which is used for such cases:

```
MWWorld::Ptr weaponPtr = *weapon;
```

For some reason we treat empty pointer as an invalid one here.

Reply:
> we've never seen that runtime_error pop up.

That's because we use early checks: 

```
    if (weapon == inv.end())
        return;

    MWWorld::Ptr weaponPtr = *weapon;
```

The point is that iterator can point to nothing rather than on item, so in this case an attempt to de-reference such iterator causes an exception.

Reply:
If we need a discussion first, I can just drop this change from this PR.

Reply:
So that last check isn't necessary in your opinion? Should we still not er on the side of caution and replace the throw with a error log entry rather than just removing the check?

Reply:
> So that last check isn't necessary in your opinion?

I do not know, this function is used in too many places. Personally I'd prefer to do not touch it.

Reply:
It's hasn't thrown "yet" in so many years and you just end up catching it and then logging an error anyway. What's the difference?

Reply:
> What's the difference?

An only difference is that I do not touch hundreds places in game mechanics, which use this function.

Reply:
I'm not asking to touch hundreds of places though. No other places where op container* operator catches the exception either.

I think it is just silly to have a try/catch in a destructor for something that has shown to not throw an exception. The only thing you do with the exception is to then ignore it anyway and throw up a log message. The only reason you're motivated to 'fix' the issue is because coverity claims it to be an issue.

If you want me to make the change, that's fine too. Revert the try/catch and I'll merge this PR in.

Reply:
> Dereference happens when you get a reference to Ptr from iterator.

I've come back around on this after looking at the code, thanks for pointing it out @akortunov